packages/mui-base/src/NumberField/NumberField.tsx

```diff
@@ -0,0 +1,200 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { NumberFieldContext } from './NumberFieldContext';
+import type { NumberFieldOwnerState, NumberFieldProps } from './NumberField.types';
+import { resolveClassName } from '../utils/resolveClassName';
+import { useNumberField } from '../useNumberField/useNumberField';
+
+function defaultRender(props: React.ComponentPropsWithRef<'div'>) {
+  return <div {...props} />;
+}
+
+/**
+ * The foundation for building custom-styled number fields.
+ *
+ * Demos:
+ *
+ * - [NumberField](https://mui.com/base-ui/react-number-field/)
+ *
+ * API:
+ *
+ * - [NumberField API](https://mui.com/base-ui/react-number-field/components-api/#number-field)
+ */
+const NumberField = React.forwardRef(function NumberField(
+  props: NumberFieldProps,
+  forwardedRef: React.ForwardedRef<HTMLDivElement>,
+) {
+  const {
+    id,
+    min,
+    max,
+    smallStep,
+    step,
+    largeStep,
+    required = false,
+    disabled = false,
+    invalid = false,
+    readOnly = false,
+    autoFocus,
+    name,
+    value,
+    onChange,
+    allowWheelScrub,
+    render: renderProp,
```

Comment:
We went with the `??` fallback because it avoids requiring a default when generating docs which is too difficult with the generic `BaseUiCommonComponentProps` due to the polymorphic components.